cmd/uncloud/ps.go

```diff
@@ -0,0 +1,241 @@
+	}
+	defer client.Close()
+
+	var containers []containerInfo
+	err = spinner.New().
+		Title(" Collecting container info...").
+		Type(spinner.MiniDot).
+		Style(lipgloss.NewStyle().Foreground(lipgloss.Color("3"))).
+		ActionWithErr(func(ctx context.Context) error {
+			containers, err = collectContainers(ctx, client)
+			return err
+		}).
+		Run()
+	if err != nil {
+		return fmt.Errorf("collect containers: %w", err)
+	}
+
+	// Sort the containers based on the sorting option
+	sort.SliceStable(containers, func(i, j int) bool {
+		a, b := containers[i], containers[j]
+		switch opts.sortBy {
+		case sortByHealth:
+			if a.highlight != b.highlight {
+				return a.highlight < b.highlight
+			}
+			if a.serviceName != b.serviceName {
+				return a.serviceName < b.serviceName
+			}
+			if a.machineName != b.machineName {
+				return a.machineName < b.machineName
+			}
+		case sortByMachine:
+			if a.machineName != b.machineName {
+				return a.machineName < b.machineName
+			}
+			if a.serviceName != b.serviceName {
+				return a.serviceName < b.serviceName
+			}
+		default: // sortByService
+			if a.serviceName != b.serviceName {
+				return a.serviceName < b.serviceName
+			}
+			if a.machineName != b.machineName {
+				return a.machineName < b.machineName
+			}
+		}
+		// Final tie-breaker
+		return a.name < b.name
+	})
+
+	return printContainers(os.Stdout, containers)
+}
+
+func printContainers(out io.Writer, containers []containerInfo) error {
+	w := tabwriter.NewWriter(out, 0, 0, 3, ' ', 0)
+	defer w.Flush()
+
+	fmt.Fprintln(w, "SERVICE\tCONTAINER ID\tNAME\tIMAGE\tSTATUS\tMACHINE")
+
+	for _, ctr := range containers {
+		id := ctr.id
+		if len(id) > 12 {
+			id = id[:12]
+		}
+
+		var statusStyle lipgloss.Style
+		switch ctr.highlight {
+		case highlightSuccess:
+			statusStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("2")) // Green
+		case highlightDanger:
+			statusStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("1")) // Red
+		case highlightWarning:
+			statusStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("3")) // Yellow
+		default:
+			statusStyle = lipgloss.NewStyle() // Default
+		}
+
+		fmt.Fprintf(
+			w, "%s\t%s\t%s\t%s\t%s\t%s\n",
+			ctr.serviceName, id, ctr.name, ctr.image,
+			statusStyle.Render(ctr.status),
+			ctr.machineName,
+		)
+	}
+	return nil
+}
+
+func collectContainers(ctx context.Context, cli *client.Client) ([]containerInfo, error) {
+	machines, err := cli.ListMachines(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("list machines: %w", err)
+	}
+
+	// List all available machines and create a list of IPs and an IP to name map
+	machinesNamesByIP := make(map[string]string)
+	md := metadata.New(nil)
+	for _, m := range machines {
+		if addr, err := m.Machine.Network.ManagementIp.ToAddr(); err == nil {
+			machineIP := addr.String()
+			machinesNamesByIP[machineIP] = m.Machine.Name
+			md.Append("machines", machineIP)
+		}
+	}
+	listCtx := metadata.NewOutgoingContext(ctx, md)
```

Comment:
JFYI you can also use this helper to create a context for broadcasting: https://github.com/psviderski/uncloud/blob/abea3e229e0989b33d3adac6e54b0ae37cddb4d8/pkg/api/client.go#L62-L98

Reply:
Thanks for pointing that out! I've updated this to use `ProxyMachinesContext`.